xwayland/selection/incoming.c

```diff
@@ -25,113 +40,115 @@ static int xwm_data_source_write(int fd, uint32_t mask, void *data) {
+	if (transfer->incr && transfer->wl_client_fd < 0) {
+		// Wayland client closed its pipe prematurely before the X11 client finished
+		// its incremental transfer. Continue draining the X11 client.
+		xwm_notify_ready_for_next_incr_chunk(transfer);
+		return;
+	}
 
-	if (transfer->property_reply != NULL) {
-		struct wl_event_loop *loop =
-			wl_display_get_event_loop(xwm->xwayland->wl_display);
-		transfer->event_source = wl_event_loop_add_fd(loop,
-			transfer->wl_client_fd, WL_EVENT_WRITABLE, xwm_data_source_write,
-			transfer);
+	bool wl_client_finished_consuming = !write_selection_property_to_wl_client(
+			transfer->wl_client_fd, WL_EVENT_WRITABLE, transfer);
+	if (wl_client_finished_consuming) {
+		return;
 	}
+
+	// Wrote out part of the property to the Wayland client, but the client was
+	// unable to accept all of it. Schedule an event to asynchronously complete
+	// the transfer.
+	struct wl_event_loop *loop =
+		wl_display_get_event_loop(transfer->selection->xwm->xwayland->wl_display);
+	transfer->event_source = wl_event_loop_add_fd(loop,
+		transfer->wl_client_fd, WL_EVENT_WRITABLE,
+		write_selection_property_to_wl_client, transfer);
 }
 
-void xwm_get_incr_chunk(struct wlr_xwm_selection_transfer *transfer) {
+static bool xwm_selection_transfer_get_selection_property(
+		struct wlr_xwm_selection_transfer *transfer, bool delete) {
 	struct wlr_xwm *xwm = transfer->selection->xwm;
-	wlr_log(WLR_DEBUG, "xwm_get_incr_chunk");
 
-	xcb_get_property_cookie_t cookie = xcb_get_property(xwm->xcb_conn,
-		0, // delete
+	xcb_get_property_cookie_t cookie = xcb_get_property(
+		xwm->xcb_conn,
+		delete,
 		transfer->selection->window,
 		xwm->atoms[WL_SELECTION],
 		XCB_GET_PROPERTY_TYPE_ANY,
 		0, // offset
 		0x1fffffff // length
 		);
 
-	xcb_get_property_reply_t *reply =
+	transfer->property_start = 0;
+	transfer->property_reply =
 		xcb_get_property_reply(xwm->xcb_conn, cookie, NULL);
-	if (reply == NULL) {
+
+	if (!transfer->property_reply) {
 		wlr_log(WLR_ERROR, "cannot get selection property");
+		return false;
+	}
+
+	return true;
+}
+
+void xwm_get_incr_chunk(struct wlr_xwm_selection_transfer *transfer) {
+	wlr_log(WLR_DEBUG, "xwm_get_incr_chunk");
+
+	if (transfer->property_reply) {
+		wlr_log(WLR_ERROR, "X11 client offered a new property before we deleted");
+		return;
+	}
+
+	if (!xwm_selection_transfer_get_selection_property(transfer, false)) {
 		return;
 	}
-	//dump_property(xwm, xwm->atoms[WL_SELECTION], reply);
 
-	if (xcb_get_property_value_length(reply) > 0) {
-		/* Reply's ownership is transferred to xwm, which is responsible
-		 * for freeing it */
-		xwm_write_property(transfer, reply);
+	if (xcb_get_property_value_length(transfer->property_reply) > 0) {
+		xwm_write_selection_property_to_wl_client(transfer);
 	} else {
-		wlr_log(WLR_DEBUG, "transfer complete");
-		xwm_selection_transfer_close_wl_client_fd(transfer);
-		free(reply);
+		xwm_selection_transfer_finished(transfer);
```

Comment:
Note for reviewer: if we hit this branch, in principle the event source has already been destroyed. However, we use `xwm_selection_transfer_finished`, which has an internal "useless" call to `xwm_selection_transfer_remove_event_source` so future readers don't have to figure this out :)